### PR TITLE
Sales analyst

### PR DESCRIPTION
**Pull request by carolinectan**
What We Did:
-Linked sales analyst with sales engine
-Created sales analyst tests and methods

Roadblocks:
-Standard deviation denominator

Tests:
-RSpec tests passing
-Spec harness tests, except for 3-4 related to BigDecimal

Notes:
-Need to update tests for fixture files after merging iteration0_specharness_refactor branch

### CODE REVIEW

File: spec/sales_engine_spec.rb [hunk 1]
```diff
@@ -26,4 +27,14 @@
     expect(se.items).to be_an_instance_of(ItemRepository)
     expect(se.merchants).to be_an_instance_of(MerchantRepository)
   end
+
+  it 'can initialize sales analyst' do
+    se = SalesEngine.from_csv({
+      :items     => "./data/items.csv",
+      :merchants => "./data/merchants.csv"
+    })
```
Comment (carolinectan) [inline spec/sales_engine_spec.rb]: Updated indentation

File: spec/sales_engine_spec.rb [hunk 2]
```diff
@@ -26,4 +27,14 @@
     expect(se.items).to be_an_instance_of(ItemRepository)
     expect(se.merchants).to be_an_instance_of(MerchantRepository)
   end
+
+  it 'can initialize sales analyst' do
+    se = SalesEngine.from_csv({
+      :items     => "./data/items.csv",
+      :merchants => "./data/merchants.csv"
```
Comment (carolinectan) [inline spec/sales_engine_spec.rb]: Updated indentation

File: lib/sales_analyst.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      (num-average_price_per_item)**2
+    end
+    denominator = (item_price_set.length - 1).to_f
+    std_dev = Math.sqrt(numerator / denominator).round(2)
+  end
+
+  def golden_items
+    @engine.items.all.select do |item|
+      item.unit_price_to_dollars > (average_price_per_item +
+        (average_price_per_item_standard_deviation * 2)) # 28 items > 13.5 mean + 10.86 std dev times 2
```
Comment (carolinectan) [inline lib/sales_analyst.rb]: Names of methods are inherently long. Ok.

File: lib/sales_analyst.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  def average_items_per_merchant_standard_deviation
+    numerator = items_by_merch_count.sum do |num|
+      (num-average_items_per_merchant)**2
+    end
+    denominator = (items_by_merch_count.length - 1).to_f
+    std_dev = Math.sqrt(numerator / denominator).round(2)
+  end
+
+  def merchants_with_high_item_count
+    merch_high_count = merch_items_hash.select do |merch_id, items|
```
[the remark this answers is not in the document]
Comment (carolinectan) [reply lib/sales_analyst.rb:36]: Ignore.

File: lib/sales_analyst.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      item.count
+    end
+  end
+
+  def average_items_per_merchant_standard_deviation
+    numerator = items_by_merch_count.sum do |num|
+      (num-average_items_per_merchant)**2
+    end
+    denominator = (items_by_merch_count.length - 1).to_f
+    std_dev = Math.sqrt(numerator / denominator).round(2)
```
Comment (carolinectan) [inline lib/sales_analyst.rb]: Removed std_dev.

File: lib/sales_analyst.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+  def items_by_merch_count
+    merch_items_hash.values.map do |item|
+      item.count
+    end
+  end
+
+  def average_items_per_merchant_standard_deviation
+    numerator = items_by_merch_count.sum do |num|
+      (num-average_items_per_merchant)**2
```
Comment (carolinectan) [inline lib/sales_analyst.rb]: Added spaces

File: lib/sales_analyst.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+    @engine.merchants.all.map do |merchant|
+      merch_items[merchant.id] = @engine.items.find_all_by_merchant_id(merchant.id)
+    end # could use merchant as the key instead of the ID in case we need to call upon the object in the future?
+
+    merch_items
+  end
+
+  def items_by_merch_count
+    merch_items_hash.values.map do |item|
```
[the remark this answers is not in the document]
Comment (carolinectan) [reply lib/sales_analyst.rb:22]: Ignore for readability

File: lib/sales_analyst.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  def average_items_per_merchant
+    (@engine.items.all.length / @engine.merchants.all.length.to_f).round(2)
+  end
+
+  def merch_items_hash
+    merch_items = {}
+
+    @engine.merchants.all.map do |merchant|
+      merch_items[merchant.id] = @engine.items.find_all_by_merchant_id(merchant.id)
+    end # could use merchant as the key instead of the ID in case we need to call upon the object in the future?
```
Comment (carolinectan) [inline lib/sales_analyst.rb]: Comment moved

File: lib/sales_analyst.rb [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    end
+  end
+
+  def average_price_per_item
+    (item_price_set.sum / item_price_set.length).round(2)
+  end
+
+  def average_price_per_item_standard_deviation
+    numerator = item_price_set.sum do |num|
+      (num-average_price_per_item)**2
```
Comment (antoniojking) [inline lib/sales_analyst.rb]: fix spacing around minus sign

File: lib/sales_analyst.rb [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+  def average_average_price_per_merchant
+    avg_price_per_merch = merch_items_hash.keys.map do |id|
+      average_item_price_for_merchant(id)
+    end
+    (avg_price_per_merch.sum / avg_price_per_merch.length).round(2)
+  end
+
+  def item_price_set
+    @engine.items.all.map do |item|
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:62]: code formatting is preferred style and syntax, ignore

File: lib/sales_analyst.rb [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  def price_of_items_for_merch(merchant_id)
+    merch_items_hash[merchant_id].map do |item|
+      item.unit_price_to_dollars
+    end
+  end
+
+  def average_item_price_for_merchant(merchant_id)
+    price_of_items_for_merch(merchant_id).sum / price_of_items_for_merch(merchant_id).length
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:51]: consider shorter variables in the future.

File: lib/sales_analyst.rb [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    merch_high_count = merch_items_hash.select do |merch_id, items|
+      items.length > (average_items_per_merchant + average_items_per_merchant_standard_deviation) # 6 items > 1.33 mean + .58 std dev
+    end
+    merch_high_count.keys.map do |merch_id|
+      @engine.merchants.find_by_id(merch_id)
+    end
+  end
+
+  def price_of_items_for_merch(merchant_id)
+    merch_items_hash[merchant_id].map do |item|
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:45]: code formatting and syntax preferred as is; ignore.

File: lib/sales_analyst.rb [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    numerator = items_by_merch_count.sum do |num|
+      (num-average_items_per_merchant)**2
+    end
+    denominator = (items_by_merch_count.length - 1).to_f
+    std_dev = Math.sqrt(numerator / denominator).round(2)
+  end
+
+  def merchants_with_high_item_count
+    merch_high_count = merch_items_hash.select do |merch_id, items|
+      items.length > (average_items_per_merchant + average_items_per_merchant_standard_deviation) # 6 items > 1.33 mean + .58 std dev
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:37]: consider shorter variables in the future; ignore.

File: lib/item.rb [hunk 14]
```diff
@@ -9,13 +9,21 @@ class Item
...
     @updated_at  = data[:updated_at]
-    @merchant_id = data[:merchant_id]
+    @merchant_id = data[:merchant_id].to_i
+  end
+
+  def update(id, attributes)
+    return @name = attributes[:name] if attributes[:name] != nil
+
+    return @description = attributes[:description] if attributes[:description] != nil
+    return @unit_price  = (attributes[:unit_price]) if attributes[:unit_price] != nil
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/item.rb:25]: will be corrected with merge of iteration0_specharness_refactor

File: lib/item.rb [hunk 15]
```diff
@@ -9,13 +9,21 @@ class Item
...
     @created_at  = data[:created_at]
     @updated_at  = data[:updated_at]
-    @merchant_id = data[:merchant_id]
+    @merchant_id = data[:merchant_id].to_i
+  end
+
+  def update(id, attributes)
+    return @name = attributes[:name] if attributes[:name] != nil
+
+    return @description = attributes[:description] if attributes[:description] != nil
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/item.rb:24]: will be corrected with merge of iteration0_specharness_refactor

File: lib/item.rb [hunk 16]
```diff
@@ -9,13 +9,21 @@ class Item
...
-    @unit_price  = data[:unit_price]
+    @unit_price  = BigDecimal(data[:unit_price])
     @created_at  = data[:created_at]
     @updated_at  = data[:updated_at]
-    @merchant_id = data[:merchant_id]
+    @merchant_id = data[:merchant_id].to_i
+  end
+
+  def update(id, attributes)
+    return @name = attributes[:name] if attributes[:name] != nil
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/item.rb:22]: will be corrected with merge of iteration0_specharness_refactor

File: lib/item.rb [hunk 17]
```diff
@@ -9,13 +9,21 @@ class Item
...
     @description = data[:description]
-    @unit_price  = data[:unit_price]
+    @unit_price  = BigDecimal(data[:unit_price])
     @created_at  = data[:created_at]
     @updated_at  = data[:updated_at]
-    @merchant_id = data[:merchant_id]
+    @merchant_id = data[:merchant_id].to_i
+  end
+
+  def update(id, attributes)
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/item.rb:21]: ignore

File: lib/merchant_repository.rb [hunk 18]
```diff
@@ -31,28 +28,30 @@ def find_by_id(id)
...
-    merchant = Merchant.new({:id => new_id.id + 1, :name => attribute})
+    merchant = Merchant.new({:id => new_id.id + 1, :name => attribute[:name]})
     @all << merchant
     merchant
   end
 
   def update(id, new_name)
     name_edit = find_by_id(id)
-    name_edit.name = new_name
+    if name_edit != nil
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/merchant_repository.rb:52]: preferred code formatting and syntax as is; ignore

File: lib/sales_analyst.rb [hunk 19]
```diff
@@ -0,0 +1,86 @@
+class SalesAnalyst
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:1]: ignore

File: spec/item_repository_spec.rb [hunk 20]
```diff
@@ -93,29 +97,29 @@
...
-      :name        => "Gold Fountain Pen",
+      # :name        => "Gold Fountain Pen",
       :description => "You can write REALLY fancy things",
       :unit_price  => 12009,
     }
     ir.update(3, data)
     expected = ir.find_by_id(3)
 
-    expect(expected.name).to eq("Gold Fountain Pen")
+    expect(expected.name).to eq("Marker")
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/item_repository_spec.rb:114]: ignore

File: spec/merchant_repository_spec.rb [hunk 21]
```diff
@@ -57,10 +59,12 @@
...
   it 'finds merchant by id and updates name' do
     mr = MerchantRepository.new("./spec/fixture_files/merchant_fixture.csv")
-    mr.update(5, "Turing School of Coding")
+    attributes = {:name => "Another Merchant"}
+
+    mr.update(5, attributes)
     expected = mr.find_by_id(5)
 
-    expect(expected.name).to eq("Turing School of Coding")
+    expect(expected.name).to eq("Another Merchant")
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/merchant_repository_spec.rb:67]: ignore

File: lib/merchant_repository.rb [hunk 22]
```diff
@@ -31,28 +28,30 @@ def find_by_id(id)
...
+      merchant.name.downcase.include?(name.downcase)
     end
   end
 
   def create(attribute)
     new_id = @all.max_by do |merchant|
       merchant.id
     end
-    merchant = Merchant.new({:id => new_id.id + 1, :name => attribute})
+    merchant = Merchant.new({:id => new_id.id + 1, :name => attribute[:name]})
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/merchant_repository.rb:45]: ignore

File: lib/sales_analyst.rb [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    end
+  end
+
+  def average_price_per_item
+    (item_price_set.sum / item_price_set.length).round(2)
+  end
+
+  def average_price_per_item_standard_deviation
+    numerator = item_price_set.sum do |num|
+      (num-average_price_per_item) ** 2
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb]: resolved

File: lib/sales_analyst.rb [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+
+  def items_by_merch_count
+    merch_items_hash.values.map do |item|
+      item.count
+    end
+  end
+
+  def average_items_per_merchant_standard_deviation
+    numerator = items_by_merch_count.sum do |num|
+      (num-average_items_per_merchant) ** 2
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb]: resolved

File: lib/sales_analyst.rb [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+
+  def average_items_per_merchant
+    (@engine.items.all.length / @engine.merchants.all.length.to_f).round(2)
+  end
+
+  def merch_items_hash
+    merch_items = {}
+    # could use merchant as the key instead of the ID in case we need to call upon the object in the future?
+    @engine.merchants.all.map do |merchant|
+      merch_items[merchant.id] = @engine.items.find_all_by_merchant_id(merchant.id)
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:15]: ignore

File: lib/sales_analyst.rb [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+    @engine = engine
+  end
+
+  def average_items_per_merchant
+    (@engine.items.all.length / @engine.merchants.all.length.to_f).round(2)
+  end
+
+  def merch_items_hash
+    merch_items = {}
+    # could use merchant as the key instead of the ID in case we need to call upon the object in the future?
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:13]: ignore

File: lib/sales_analyst.rb [hunk 27]
```diff
@@ -0,0 +1,86 @@
+class SalesAnalyst
+
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply lib/sales_analyst.rb:2]: ignore

File: spec/merchant_repository_spec.rb [hunk 28]
```diff
@@ -57,10 +59,12 @@
 
   it 'finds merchant by id and updates name' do
     mr = MerchantRepository.new("./spec/fixture_files/merchant_fixture.csv")
-    mr.update(5, "Turing School of Coding")
+    attributes = {:name => "Another Merchant"}
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/merchant_repository_spec.rb:62]: ignore

File: spec/merchant_repository_spec.rb [hunk 29]
```diff
@@ -48,7 +48,9 @@
 
   it 'creates a new merchant with attributes' do
     mr = MerchantRepository.new("./spec/fixture_files/merchant_fixture.csv")
-    mr.create("Another Merchant")
+    attributes = {:name => "Another Merchant"}
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/merchant_repository_spec.rb:51]: ignore

File: spec/sales_analyst_spec.rb [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+  it 'can average prices of items' do
+    expect(@sales_analyst.average_price_per_item).to eq(13.50)
+  end
+
+  it 'can average prices of items by standard deviation' do
+    expect(@sales_analyst.average_price_per_item_standard_deviation).to eq(5.43)
+  end
+
+  it 'can return golden items' do
+    expect(@sales_analyst.golden_items).to eq([]) # => [<item>, <item>, <item>, <item>]
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:77]: ignore

File: spec/sales_analyst_spec.rb [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+  it 'can return the average item price for a specific merchant' do
+    expect(@sales_analyst.price_of_items_for_merch(7)).to eq([12.0, 20.0]) # => BigDecimal
+  end
+
+  it 'can return the average item price for a specific merchant' do
+    expect(@sales_analyst.average_item_price_for_merchant(7)).to eq(16) # => BigDecimal
+  end
+
+  it 'can return the average price per merchant' do
+    expect(@sales_analyst.average_average_price_per_merchant).to eq(12.67) # => BigDecimal
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:61]: ignore

File: spec/sales_analyst_spec.rb [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+    expect(@sales_analyst.merchants_with_high_item_count.length).to eq(1)
+    expect(@sales_analyst.merchants_with_high_item_count).to eq([@se.merchants.find_by_id(7)])
+  end
+
+  it 'can return the average item price for a specific merchant' do
+    expect(@sales_analyst.price_of_items_for_merch(7)).to eq([12.0, 20.0]) # => BigDecimal
+  end
+
+  it 'can return the average item price for a specific merchant' do
+    expect(@sales_analyst.average_item_price_for_merchant(7)).to eq(16) # => BigDecimal
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:57]: ignore

File: spec/sales_analyst_spec.rb [hunk 33]
```diff
@@ -0,0 +1,89 @@
...
+    expect(@sales_analyst.average_items_per_merchant_standard_deviation).to eq(0.58)
+  end
+
+  it 'can return merchants with high item count' do
+    expect(@sales_analyst.merchants_with_high_item_count.length).to eq(1)
+    expect(@sales_analyst.merchants_with_high_item_count).to eq([@se.merchants.find_by_id(7)])
+  end
+
+  it 'can return the average item price for a specific merchant' do
+    expect(@sales_analyst.price_of_items_for_merch(7)).to eq([12.0, 20.0]) # => BigDecimal
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:53]: ignore

File: spec/sales_analyst_spec.rb [hunk 34]
```diff
@@ -0,0 +1,89 @@
...
+    # std_dev = sqrt( ( (1-1.33)^2+(1-1.33)^2+(2-1.33)^2 ) / 2 )
+    # std_dev = sqrt( ( .6667 / 2 )
+    # std_dev = sqrt( ( .33335 )
+    # std_dev = .578
+    expect(@sales_analyst.average_items_per_merchant_standard_deviation).to eq(0.58)
+  end
+
+  it 'can return merchants with high item count' do
+    expect(@sales_analyst.merchants_with_high_item_count.length).to eq(1)
+    expect(@sales_analyst.merchants_with_high_item_count).to eq([@se.merchants.find_by_id(7)])
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:49]: ignore

File: spec/sales_analyst_spec.rb [hunk 35]
```diff
@@ -0,0 +1,89 @@
...
+  end
+
+  it 'can return average items per merchant standard deviation' do
+    # 1.33 is the mean
+    # set = [1, 1, 2]
+    # std_dev = sqrt( ( (1-1.33)^2+(1-1.33)^2+(2-1.33)^2 ) / 2 )
+    # std_dev = sqrt( ( .6667 / 2 )
+    # std_dev = sqrt( ( .33335 )
+    # std_dev = .578
+    expect(@sales_analyst.average_items_per_merchant_standard_deviation).to eq(0.58)
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:44]: ignore

File: spec/sales_analyst_spec.rb [hunk 36]
```diff
@@ -0,0 +1,89 @@
...
+    expected = {
+                5 => @se.items.find_all_by_merchant_id(5),
+                6 => @se.items.find_all_by_merchant_id(6),
+                7 => @se.items.find_all_by_merchant_id(7)
+               }
+    expect(@sales_analyst.merch_items_hash).to eq(expected)
+  end
+
+  it 'can return items by merchant count' do
+    expect(@sales_analyst.items_by_merch_count).to eq([1,1,2])
```
Comment (antoniojking) [inline spec/sales_analyst_spec.rb]: fix spacing

File: spec/sales_analyst_spec.rb [hunk 37]
```diff
@@ -0,0 +1,89 @@
...
+  it 'can return average items per merchant' do
+    expect(@sales_analyst.average_items_per_merchant).to eq(1.33)
+  end
+
+  it 'can return a hash of the merchant ids and items' do
+    expected = {
+                5 => @se.items.find_all_by_merchant_id(5),
+                6 => @se.items.find_all_by_merchant_id(6),
+                7 => @se.items.find_all_by_merchant_id(7)
+               }
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:29]: ignore

File: spec/sales_analyst_spec.rb [hunk 38]
```diff
@@ -0,0 +1,89 @@
...
+    expect(@sales_analyst).to be_an_instance_of(SalesAnalyst)
+  end
+
+  it 'can return average items per merchant' do
+    expect(@sales_analyst.average_items_per_merchant).to eq(1.33)
+  end
+
+  it 'can return a hash of the merchant ids and items' do
+    expected = {
+                5 => @se.items.find_all_by_merchant_id(5),
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:26]: ignore

File: spec/sales_analyst_spec.rb [hunk 39]
```diff
@@ -0,0 +1,89 @@
...
+
+require_relative '../lib/sales_engine'
+require_relative '../lib/sales_analyst'
+
+RSpec.describe SalesEngine do
+  before :each do
+    @se = SalesEngine.from_csv({
+                                :items => './spec/fixture_files/item_fixture.csv',
+                                :merchants => './spec/fixture_files/merchant_fixture.csv'
+                              })
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:12]: ignore

File: spec/sales_analyst_spec.rb [hunk 40]
```diff
@@ -0,0 +1,89 @@
...
+SimpleCov.start
+
+require_relative '../lib/sales_engine'
+require_relative '../lib/sales_analyst'
+
+RSpec.describe SalesEngine do
+  before :each do
+    @se = SalesEngine.from_csv({
+                                :items => './spec/fixture_files/item_fixture.csv',
+                                :merchants => './spec/fixture_files/merchant_fixture.csv'
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:11]: ignore

File: spec/sales_analyst_spec.rb [hunk 41]
```diff
@@ -0,0 +1,89 @@
+require 'SimpleCov'
+SimpleCov.start
+
+require_relative '../lib/sales_engine'
+require_relative '../lib/sales_analyst'
+
+RSpec.describe SalesEngine do
+  before :each do
+    @se = SalesEngine.from_csv({
+                                :items => './spec/fixture_files/item_fixture.csv',
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:10]: ignore

File: spec/sales_analyst_spec.rb [hunk 42]
```diff
@@ -0,0 +1,89 @@
+require 'SimpleCov'
+SimpleCov.start
+
+require_relative '../lib/sales_engine'
+require_relative '../lib/sales_analyst'
+
+RSpec.describe SalesEngine do
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:7]: ignore

File: spec/sales_analyst_spec.rb [hunk 43]
```diff
@@ -0,0 +1,89 @@
+require 'SimpleCov'
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:1]: ignore

File: spec/sales_engine_spec.rb [hunk 44]
```diff
@@ -4,26 +4,37 @@
...
 
     expect(se.items).to be_an_instance_of(ItemRepository)
     expect(se.merchants).to be_an_instance_of(MerchantRepository)
   end
+
+  it 'can initialize sales analyst' do
+    se = SalesEngine.from_csv({
+                               :items     => "./data/items.csv",
+                               :merchants => "./data/merchants.csv"
+                             })
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:35]: ignore

File: spec/sales_engine_spec.rb [hunk 45]
```diff
@@ -4,26 +4,37 @@
...
     se = SalesEngine.from_csv(data)
 
     expect(se.items).to be_an_instance_of(ItemRepository)
     expect(se.merchants).to be_an_instance_of(MerchantRepository)
   end
+
+  it 'can initialize sales analyst' do
+    se = SalesEngine.from_csv({
+                               :items     => "./data/items.csv",
+                               :merchants => "./data/merchants.csv"
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:34]: ignore

File: spec/sales_engine_spec.rb [hunk 46]
```diff
@@ -4,26 +4,37 @@
...
 
     se = SalesEngine.from_csv(data)
 
     expect(se.items).to be_an_instance_of(ItemRepository)
     expect(se.merchants).to be_an_instance_of(MerchantRepository)
   end
+
+  it 'can initialize sales analyst' do
+    se = SalesEngine.from_csv({
+                               :items     => "./data/items.csv",
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:33]: ignore

File: spec/sales_engine_spec.rb [hunk 47]
```diff
@@ -4,26 +4,37 @@
...
   end
 
   it 'initializes with attributes that create repository instances' do
     data = {
-      :items     => "./data/items.csv",
-      :merchants => "./data/merchants.csv"
-    }
+            :items     => "./data/items.csv",
+            :merchants => "./data/merchants.csv"
+           }
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:23]: ignore

File: spec/sales_engine_spec.rb [hunk 48]
```diff
@@ -4,26 +4,37 @@
...
     expect(se).to be_an_instance_of(SalesEngine)
   end
 
   it 'initializes with attributes that create repository instances' do
     data = {
-      :items     => "./data/items.csv",
-      :merchants => "./data/merchants.csv"
-    }
+            :items     => "./data/items.csv",
+            :merchants => "./data/merchants.csv"
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:22]: ignore

File: spec/sales_engine_spec.rb [hunk 49]
```diff
@@ -4,26 +4,37 @@
...
 
     expect(se).to be_an_instance_of(SalesEngine)
   end
 
   it 'initializes with attributes that create repository instances' do
     data = {
-      :items     => "./data/items.csv",
-      :merchants => "./data/merchants.csv"
-    }
+            :items     => "./data/items.csv",
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:21]: ignore

File: spec/sales_engine_spec.rb [hunk 50]
```diff
@@ -4,26 +4,37 @@
...
 
 RSpec.describe SalesEngine do
   it 'exists' do
     se = SalesEngine.from_csv({
-      :items     => "./data/items.csv",
-      :merchants => "./data/merchants.csv"
-    })
+                               :items     => "./data/items.csv",
+                               :merchants => "./data/merchants.csv"
+                             })
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:14]: ignore

File: spec/sales_engine_spec.rb [hunk 51]
```diff
@@ -4,26 +4,37 @@
...
+require_relative '../lib/sales_analyst'
 
 RSpec.describe SalesEngine do
   it 'exists' do
     se = SalesEngine.from_csv({
-      :items     => "./data/items.csv",
-      :merchants => "./data/merchants.csv"
-    })
+                               :items     => "./data/items.csv",
+                               :merchants => "./data/merchants.csv"
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:13]: ignore

File: spec/sales_engine_spec.rb [hunk 52]
```diff
@@ -4,26 +4,37 @@
...
 require_relative '../lib/merchant_repository'
+require_relative '../lib/sales_analyst'
 
 RSpec.describe SalesEngine do
   it 'exists' do
     se = SalesEngine.from_csv({
-      :items     => "./data/items.csv",
-      :merchants => "./data/merchants.csv"
-    })
+                               :items     => "./data/items.csv",
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_engine_spec.rb:12]: ignore

File: spec/sales_analyst_spec.rb [hunk 53]
```diff
@@ -0,0 +1,89 @@
...
+SimpleCov.start
+
+require_relative '../lib/sales_engine'
+require_relative '../lib/sales_analyst'
+
+RSpec.describe SalesEngine do
+  before :each do
+    @se = SalesEngine.from_csv({
```
[the remark this answers is not in the document]
Comment (antoniojking) [reply spec/sales_analyst_spec.rb:9]: ignore

**Comment by antoniojking** [comment]
Looks good! No merge conflicts. Let's work on correcting spacing in hashes and arrays, proper indentation of hashes, using single quotes when interpolation is not required, and alternatives for !nil.

**Pull request merged**